File: Programmers/단어변환.cpp
#include <string>
#include <vector>

using namespace std;
int answer = 51;
bool visit[50];

void DFS(string begin, string target, vector<string> words, int depth) {

	if (begin == target) {
		answer = answer > depth ? depth : answer;
		return;
	}
	for (int i = 0; i<words.size(); i++) {
		int count = 0;
		if (!visit[i]) {
			for (int j = 0; j<words[i].length(); j++) {
				if (begin[j] != words[i][j]) {
					count++;
					if (count == 2) break;
				}
			}
			if (count != 2) {
				visit[i] = true;
				DFS(words[i], target, words, depth + 1);
				visit[i] = false;
			}
		}
	}
}

int solution(string begin, string target, vector<string> words) {

	DFS(begin, target, words, 0);

	if (answer == 51) answer = 0;
	return answer;
} 

File: Programmers/단어변환.java
class Solution {
    
    boolean []visit = new boolean[50];
    int answer = 51;
    void DFS(String begin, String target, String[] words, int depth){
        if(begin.equals(target)){
            answer = answer > depth ? depth : answer;
            return;
        }
        for(int i=0; i<words.length; i++){
            if(!visit[i]){
                int diff = 0;
                for(int k=0; k<words[i].length(); k++){
                    if(begin.charAt(k) != words[i].charAt(k)){
                        diff++;
                    }
                    if(diff == 2) break;
                }
                if(diff == 1){
                    visit[i] = true;
                    //System.out.println(words[i]);
                    DFS(words[i], target, words, depth+1);
                    visit[i] = false;
                }
            }
        }
        
    }
    
    public int solution(String begin, String target, String[] words) {
        
        String a = "a";
        String b= "b";
        if(a.compareTo(b) > 0) System.out.println("a");
        else System.out.println("b");
        
        DFS(begin, target, words, 0);
        if(answer == 51) answer = 0;
        return answer;
    }
}
